File: 704-binary-search/704-binary-search.js
// T.C. = O(logN) for how many times N(length of nums) can be indived by 2
// S.C. = O(1)

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let l = 0;
    let r = nums.length - 1;
    
    while (l <= r) {
        let m = Math.floor(l + (r - l) / 2);
        if (nums[m] < target) {
            l = m + 1;
        } else if (nums[m] > target) {
            r = m - 1;
        } else {
            return m;
        }
    }
    
    return -1;
};

File: 704-binary-search/704-binary-search.py
# Time Complexity: O(logN)
# n is the length of nums, while loop will run logN time

# Space Complexity: O(1)
# the space required by the algorithm to process data is constant

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = 0
        r = len(nums) - 1
        
        while l <= r:
            m = l + (r - l) // 2
            if nums[m] < target:
                l = m + 1
            elif nums[m] > target:
                r = m - 1
            else:
                return m
        
        return -1
        
        